### PR TITLE
New Demo Mode

### LINKED ISSUE #13: demo mode
R has the concept of a demo: https://stat.ethz.ch/R-manual/R-devel/library/utils/html/demo.html

It would be nice if we had a `demo` function, which would take a module/function and turn it into jupyter notebook cells:

```
from ipyext.demo import notebook_demo as demo # make it cristall clear that we only work in the notebook
import matplotlib.demo # something which isn't there yet :-)
demo(matplotlib.demo)
# shows available demos, as that's a module
demo(matplotlib.demo.gridspec)
# would add some cells below which are prefilled with examples how to use gridspec
```

Demo would:
- make sure that the function takes no argument
- get the source of a function and dedent it
- split the source into blocks (code  + comments with `^# Text` -> comments with `^#text` stay as comments in the code cell)
- remove `#` from all comment lines
- add the blocks to the notebook (see [here](https://github.com/jupyter-incubator/contentmanagement/blob/master/urth/cms/inject.py) for an example how to do that

Only problem when using functions: we can't use `%magics` in demos. Not sure if there is a workaround that, maybe treat `#%magic ...` as code, aka `s/#%/%/` before splitting into blocks?

### PR DESCRIPTION
**Pull request by jankatins**
Builds new notebook cells from the source of a function:

Example:

```
from ipyext.demo import demo
import ipyext.demo
# ---
demo(ipyext.demo)
# ---
demo(ipyext.demo.demo_example)
```

Closes: https://github.com/ipython-contrib/IPython-extensions/issues/13

**Todo**
- [x] write some tests for the ipython frontend
- [x] `demo("<gh_matplotlib>/")` -> prints the wrong thing...
- [x] write docs...
- [ ] How to format github examples to get them split into multiple cells? Look at some more matplotlib examples...
- [x] Use the events: `get_ipython().events.register("post_run_cell", self.method)`, see here:  https://github.com/FrancescAlted/ipython_memwatcher/blob/master/ipython_memwatcher/memwatcher.py#L47

### CODE REVIEW

**Comment by jankatins** [comment]
And now I have it working in the notebook and in ipython/qtconsole directly...

Also added matplotlib examples directly from github...

**Comment by juhasch** [comment]
Nice!
Seems like `notebook_demo` is actutally `demo` ?

``` Python
import ipyext.demo
from ipyext.demo import demo
demo(ipyext.demo.demo_example)
```

**Comment by jankatins** [comment]
@juhasch yep... Renamed it after I got the ipython version running

**Comment by jankatins** [comment]
Example and some demo of the demo mode: http://janschulz.github.io/demo_mode_ipython.html

**Comment by juhasch** [comment]
Looks like it could be really useful.

**Comment by jankatins** [comment]
Lets see if someone uses it... Having the function here and not in ipython is a little entry barrier :-/

**Comment by jankatins** [comment]
Ok, IMO this is ready if travis gets green...

The only open todo is how to handle matplotlib examples, but I tested this and the current approach (aka not splitting it into multiple cells) is the only one which works due to the way they are written (comments with `#` and also because they expect that `plt.show()` is called in the end... It's even worse when you use `%matplotlib inline` as that produces one plot per cell :-()

If it is merged I will write demos for the current magics and then lets see if there are some other projects which want to write demos...

**Comment by jankatins** [comment]
@juhasch a short review and if satisfied a merge would be nice :-)

**Comment by juhasch** [comment]
Hmm,
tried to run tests and generate docs:

```
$ make
python autogen_api.py
python: can't open file 'autogen_api.py': [Errno 2] No such file or directory
make: *** [source/api/generated/gen.txt] Error 2

$ iptest3 ipyext

Test group: ipyext
E
======================================================================
ERROR: Failure: ImportError (No module named 'ipyext')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/users/juhasch/anaconda/lib/python3.5/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/users/juhasch/anaconda/lib/python3.5/site-packages/nose/loader.py", line 407, in loadTestsFromName
    module = resolve_name(addr.module)
  File "/users/juhasch/anaconda/lib/python3.5/site-packages/nose/util.py", line 312, in resolve_name
    module = __import__('.'.join(parts_copy))
ImportError: No module named 'ipyext'

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (errors=1)

______________________________________________________________________
Test suite completed for system with the following information:
IPython version: 4.0.0
IPython commit : f534027 (installation)
IPython package: /users/juhasch/anaconda/lib/python3.5/site-packages/IPython
Python version : 3.5.0 |Anaconda 2.4.0 (x86_64)| (default, Oct 20 2015, 14:39:26) [GCC 4.2.1 (Apple Inc. build 5577)]
sys.executable : /users/juhasch/anaconda/bin/python
Platform       : Darwin-15.2.0-x86_64-i386-64bit

Tools and libraries available at test time:
   matplotlib pygments sqlite3

Status: ERROR - 1 out of 1 test groups failed (ipyext). Took 1.191s.

You may wish to rerun these, with:
  iptest ipyext
```

**Comment by jankatins** [comment]
The first (make) should be fixed (I use make.cmd and RTD something else, so it wasn't catched,,,)

**Comment by jankatins** [comment]
I can't reproduce the second problem. Could it be that you ran the tests in the docs subdir?

**Comment by juhasch** [comment]
Adding the extension to `PYTHONPATH` let's me run the tests, but now they loop endlessly.
Will look into it tomorrow.

The doc is now generated fine.

**Comment by jankatins** [comment]
there is also `python tests.py` but that should actualy be mostly the same as `iptest ipyext`.

File: ipyext/demo.py [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+    # and insert it into the cell as code.
+    #%%time
+    _sum = 0
+    for x in range(10000):
+        _sum += x
+    # Print the sum:
+    print(_sum)
+
+# This lets the `demo(ipyext.demo)` find only the `demo_example`.
+# Only modues with that variable will display an overview of
```
Comment (juhasch) [inline ipyext/demo.py]: Type: modues->modules

**Comment by juhasch** [comment]
Works fine in the notebook.
In the shell I get

``` Python
In [4]: demo(ipyext.demo.demo_example)
<IPython.core.display.Javascript object>
```

Would it make sense to return a text representation in this case ?

**Comment by jankatins** [comment]
> In [4]: demo(ipyext.demo.demo_example)
> IPython.core.display.Javascript object
> Would it make sense to return a text representation in this case ?

You need to use `demo(..., frontend="ipython")`... I've actually no idea how to find out if a kernel is run in a notebook or in the qtconsole/console :-(

**Comment by juhasch** [comment]
How about adding a `__repr__()` function ?
This should be called if `_repr_javascript_` is not supported at the output.

**Comment by jankatins** [comment]
Currently I use the normal `JavaScript` object... Hm, I would need to subclass that... ok, good idea, incoming... :-)

**Comment by jankatins** [comment]
done :-)

**Pull request closed without merging**